Get rid of bounds check in slice::chunks_exact() and related function…

sdroege opened the pull request:
…s during construction

LLVM can't figure out in

    let rem = self.len() % chunk_size;
    let len = self.len() - rem;
    let (fst, snd) = self.split_at(len);

and

    let rem = self.len() % chunk_size;
    let (fst, snd) = self.split_at(rem);

that the index passed to split_at() is smaller than the slice length and
adds a bounds check plus panic for it.

Apart from removing the overhead of the bounds check this also allows
LLVM to optimize code around the ChunksExact iterator better.

sdroege committed:
Get rid of bounds check in slice::chunks_exact() and related functions during construction

LLVM can't figure out in

    let rem = self.len() % chunk_size;
    let len = self.len() - rem;
    let (fst, snd) = self.split_at(len);

and

    let rem = self.len() % chunk_size;
    let (fst, snd) = self.split_at(rem);

that the index passed to split_at() is smaller than the slice length and
adds a bounds check plus panic for it.

Apart from removing the overhead of the bounds check this also allows
LLVM to optimize code around the ChunksExact iterator better.

sdroege commented:
See https://rust.godbolt.org/z/8P74zY for the corresponding assembly of both variants. This is based on the code from https://github.com/rust-lang/rust/issues/75935, where I noticed this. The variant without bound checks is unrolled by the compiler.

Instead of implementing `split_at`/`split_at_mut` manually, it works equally well to simply put a `core::intrinsics::assume(len < s.len())` into the code. As I didn't see that used much I didn't do it for this PR but if you prefer that I can switch it over.

Of course even better would be if LLVM would *just* understand this itself but that seems considerably harder to achieve.

sdroege commented:
As @lzutao wrote in the other issue, this might also be https://github.com/rust-lang/rust/issues/74938

sdroege commented:
Seems like #74938 makes this one here obsolete. Might still be worth merging depending on when an update to LLVM 12 is planned.

pickfire commented on library/core/src/slice/mod.rs:
```diff
@@ -865,8 +865,10 @@ impl<T> [T] {
         let rem = self.len() % chunk_size;
-        let len = self.len() - rem;
-        let (fst, snd) = self.split_at(len);
+        let fst_len = self.len() - rem;
```
Is `fst` first?

sdroege replied:
Yes, just taking the naming that was there before :)

pickfire commented:
Weird, how come there is no reviewers assigned automatically? @Dylan-DPC

matklad commented on library/core/src/slice/mod.rs:
```diff
@@ -865,8 +865,10 @@ impl<T> [T] {
+        let fst_len = self.len() - rem;
+        // The following is the same as self.split_at(fst_len) but avoids a bounds check
+        // SAFETY: 0 <= fst_len < self.len() by construction above
+        let (fst, snd) = unsafe { (self.get_unchecked(0..fst_len), self.get_unchecked(fst_len..)) };
```
Drive by comment: maybe `self.get_unchecked(..fst_len)` here and elsewhere?

pickfire replied:
Wait, I believe this is not `scd`, it should be `rem` / `rest`. `scd` is second but it does not means second here, it means all items other than first.

Dylan-DPC commented:
r? @nagisa

pickfire commented on library/core/src/slice/mod.rs:
```diff
@@ -865,8 +865,10 @@ impl<T> [T] {
+        // The following is the same as self.split_at(fst_len) but avoids a bounds check
+        // SAFETY: 0 <= fst_len < self.len() by construction above
+        let (fst, snd) = unsafe { (self.get_unchecked(0..fst_len), self.get_unchecked(fst_len..)) };
         ChunksExact { v: fst, rem: snd, chunk_size }
```
Is `snd` second? If it is why isn't it `scd`?

sdroege replied:
It is, and I don't know :) I'm fine with renaming these but that's how they were named before and also elsewhere in the same file, so if someone wants to have them renamed let's do it consistently at least everywhere in the same file.

matklad replied:
`fst` and `snd` are well-established names for these in functional programming langauges

* https://en.wikibooks.org/wiki/Haskell/Lists_and_tuples#Example:_fst_and_snd
* https://docs.racket-lang.org/plait/Predefined_Functions_and_Constants.html#%28def._%28%28lib._plait%2Fmain..rkt%29._fst%29%29
* https://www.cs.cornell.edu/courses/cs3110/2009sp/recitations/rec02.html

Amusingly, they are a more obvious alternative to [`car` and `cdr`](https://en.wikipedia.org/wiki/CAR_and_CDR)

pickfire replied:
Ah, I am not coming from haskell so I it took me a while to read the context and guess that it is `second`. I would rather do `a`, `b` and `c` to show `fst`, `snd` and `trd` (hopefully I am correct on this). Heh, thanks for explaining I think I got the pattern now. `fth`, `fth`, `sth`, `sth`, `sth`, `eth`, `nth`, `tth`...

nagisa reviewed (commented):
The only change I would suggest here is to split these inlined algorithms into a private function (`split_at_unchecked`/`split_at_mut_unchecked`), while also possibly taking advantage of them to implement the `split_at` and `split_at_mut` as well.

Your call really if you want to do this, r=me otherwise.

pickfire commented:
@nagisa Would it be good if we have `split_at_mut_unchecked` to make use of `get_unchecked` on `slice`?

nagisa commented:
Yeah, to implement the `split_at_{,mut}_unckecked` you would use `get_unchecked` much like its done right now in this PR. `split_at{,_mut}` functions, if rewritten to use the `unchecked` variants would then also be changed to contain bound-checking code.

Like I said this is not strictly necessary to land this IMO, and even if we don’t make an adjustment here, we can also do the thing in a follow-up.

sdroege commented:
I'd be happy to add the `split_at_unchecked()` here, easy enough to do so. You'd prefer that over using `core::intrinsics::assume()` here then.

sdroege committed:
Add (non-public) slice::split_at_unchecked() and split_at_mut_unchecked()

These are unsafe variants of the non-unchecked functions and don't do
any bounds checking.

For the time being these are not public and only a preparation for the
following commit. Making it public and stabilization can follow later
and be discussed in https://github.com/rust-lang/rust/issues/76014 .

sdroege commented:
`split_at_unchecked()` is added here now and made use of.

lzutao commented on library/core/src/slice/mod.rs:
```diff
@@ -865,8 +865,9 @@ impl<T> [T] {
-        let len = self.len() - rem;
-        let (fst, snd) = self.split_at(len);
+        let fst_len = self.len() - rem;
+        // SAFETY: 0 <= fst_len < self.len() by construction above
```
when `self.len() == 0` here and other places:
```suggestion
        // SAFETY: 0 <= fst_len < self.len() by construction above
```

sdroege replied:
Indeed, thanks

pickfire commented on library/core/src/slice/mod.rs:
```diff
@@ -1178,16 +1185,99 @@ impl<T> [T] {
-        }
+    /// Divides one slice into two at an index, without doing bounds checking.
+    ///
+    /// The first will contain all indices from `[0, mid)` (excluding
+    /// the index `mid` itself) and the second will contain all
+    /// indices from `[mid, len)` (excluding the index `len` itself).
+    ///
+    /// This is generally not recommended, use with caution!
+    /// Calling this method with an out-of-bounds index is *[undefined behavior]*
+    /// even if the resulting reference is not used.
+    /// For a safe alternative see [`split_at`].
```
I think this could be improved even though it is copied from `get_unchecked`, later I will probably send a patch for that.

```suggestion
    /// The first will contain all indices from `[0, mid)` (excluding
    /// the index `mid` itself) and the second will contain all
    /// indices from `[mid, len)` (excluding the index `len` itself).
    /// For a safe alternative see [`split_at`].
    ///
    /// # Safety
    ///
    /// Calling this method with an out-of-bounds index is *[undefined behavior]*
    /// even if the resulting reference is not used.
```

pickfire commented on library/core/src/slice/mod.rs:
```diff
@@ -1178,16 +1185,99 @@ impl<T> [T] {
+    /// even if the resulting reference is not used.
+    /// For a safe alternative see [`split_at`].
+    ///
+    /// [`split_at`]: #method.split_at
```
I don't think this is needed. @jyn514 would be happy if you remove it.

jyn514 replied:
This is actually on slices, so it can't be fixed with intra-doc links: #63351

pickfire commented on library/core/src/slice/mod.rs:
```diff
@@ -1178,16 +1185,99 @@ impl<T> [T] {
+    ///
+    /// unsafe {
+    ///    let (left, right) = v.split_at_unchecked(0);
+    ///    assert!(left == []);
+    ///    assert!(right == [1, 2, 3, 4, 5, 6]);
```
```suggestion
    ///    assert_eq!(left, []);
    ///    assert_eq!(right, [1, 2, 3, 4, 5, 6]);
```

pickfire commented on library/core/src/slice/mod.rs:
```diff
@@ -1178,16 +1185,99 @@ impl<T> [T] {
+    ///     let (left, right) = v.split_at_unchecked(6);
+    ///     assert!(left == [1, 2, 3, 4, 5, 6]);
+    ///     assert!(right == []);
+    /// }
```
I think we should show an example of invalid use.

pickfire commented on library/core/src/slice/mod.rs:
```diff
@@ -1178,16 +1185,99 @@ impl<T> [T] {
+    ///
+    /// unsafe {
+    ///     let (left, right) = v.split_at_unchecked(2);
+    ///     assert!(left == [1, 2]);
+    ///     assert!(right == [3, 4, 5, 6]);
```
```suggestion
    ///     assert_eq!(left, [1, 2]);
    ///     assert_eq!(right, [3, 4, 5, 6]);
```

pickfire commented on library/core/src/slice/mod.rs:
```diff
@@ -1178,16 +1185,99 @@ impl<T> [T] {
+    ///
+    /// unsafe {
+    ///     let (left, right) = v.split_at_unchecked(6);
+    ///     assert!(left == [1, 2, 3, 4, 5, 6]);
+    ///     assert!(right == []);
```
```suggestion
    ///     assert_eq!(left, [1, 2, 3, 4, 5, 6]);
    ///     assert_eq!(right, []);
```

pickfire commented on library/core/src/slice/mod.rs:
```diff
@@ -1178,16 +1185,99 @@ impl<T> [T] {
+    /// // scoped to restrict the lifetime of the borrows
+    /// unsafe {
+    ///     let (left, right) = v.split_at_mut_unchecked(2);
+    ///     assert!(left == [1, 0]);
+    ///     assert!(right == [3, 0, 5, 6]);
```
```suggestion
    ///     assert_eq!(left, [1, 0]);
    ///     assert_eq!(right, [3, 0, 5, 6]);
```

pickfire commented on library/core/src/slice/mod.rs:
```diff
@@ -1178,16 +1185,99 @@ impl<T> [T] {
+    /// The first will contain all indices from `[0, mid)` (excluding
+    /// the index `mid` itself) and the second will contain all
+    /// indices from `[mid, len)` (excluding the index `len` itself).
+    ///
```
Please update this too if the previous suggestion was applied.

pickfire commented on library/core/src/slice/mod.rs:
```diff
@@ -1178,16 +1185,99 @@ impl<T> [T] {
+    #[unstable(feature = "slice_split_at_unchecked", reason = "new API", issue = "76014")]
+    #[inline]
+    pub unsafe fn split_at_unchecked(&self, mid: usize) -> (&[T], &[T]) {
+        // SAFETY: Caller has to check that 0 <= mid < self.len()
```
We should document this in safety above, which is more important than here.

sdroege replied:
This here is needed so the CI passes. Each use of `unsafe` needs to be documented why it's OK.

All your documentation comments also apply to the non-unchecked variant and at least also `get_unchecked()`. So if they should be applied, let's do it for all of them consistently instead of keeping a mess around.

pickfire commented on library/core/src/slice/mod.rs:
```diff
@@ -1178,16 +1185,99 @@ impl<T> [T] {
+    ///     left[1] = 2;
+    ///     right[1] = 4;
+    /// }
+    /// assert!(v == [1, 2, 3, 4, 5, 6]);
```
Should we show an invalid case?

pickfire commented on library/core/src/slice/mod.rs:
```diff
@@ -1178,16 +1185,99 @@ impl<T> [T] {
+        let len = self.len();
+        let ptr = self.as_mut_ptr();
+
+        // SAFETY: Caller has to check that 0 <= mid < self.len().
```
We should document this in safety section above.

lzutao commented on library/core/src/slice/mod.rs:
```diff
@@ -1178,16 +1185,99 @@ impl<T> [T] {
+
+        // SAFETY: Caller has to check that 0 <= mid < self.len().
+        //
+        // [ptr; mid] and [mid; len] are not overlapping, so returning a mutable reference
```
```suggestion
        // `[ptr; mid]` and `[mid; len]` are not overlapping, so returning a mutable reference
```

sdroege commented:
If the general implementation is considered OK, I'll update the docs/comments accordingly on Sunday or Monday, and also the docs of `get_unchecked()` / `split_at()` in a separate commit so they stay consistent.

lzutao commented:
I would prefer to open an issue to remember to use safe function when we update LLVM 12.

sdroege commented:
> I would prefer to open an issue to remember to use safe function when we update LLVM 12.

That would still depend on the optimizer to actually figure it out in all cases, but I'll create an issue for that once this here is (ready to be) merged.

nagisa commented:
I think the `split_at_unchecked` should remain private in this PR, as I suggested initially. If getting these functions to be public is a goal, then that should probably be done in a separate PR, most important of which is that making them public would require a T-libs review and I’m not on the T-libs team. But also – private APIs don’t have requirements as stringent as a public API would.

The implementation LGTM.

sdroege committed:
Improve documentation of slice::get_unchecked() / split_at_unchecked()

Thanks to Ivan Tham, who gave the majority of these suggestions during
their review.

sdroege commented:
I've addressed all (I hope) review comments now.

> I think the `split_at_unchecked` should remain private in this PR, as I suggested initially.

Sorry I missed that in the initial comment. I've marked it as private for now so we can move ahead here, but I'd like to get them public at some point in the future so keeping the docs/etc in place. I went through various code of mine and noticed a couple of places where I could've used `split_at_mut_unchecked()` instead of the ad-hoc versions I wrote there.

nagisa commented:
@bors r+

bors commented:
:pushpin: Commit 8d3cf9237dcc9d4f2a7c0bde42dc133e60f51c7d has been approved by `nagisa`

<!-- @bors r=nagisa 8d3cf9237dcc9d4f2a7c0bde42dc133e60f51c7d -->
<!-- homu: {"type":"Approved","sha":"8d3cf9237dcc9d4f2a7c0bde42dc133e60f51c7d","approver":"nagisa"} -->

bors commented:
:hourglass: Testing commit 8d3cf9237dcc9d4f2a7c0bde42dc133e60f51c7d with merge 52d6d94eb07980426e50416d3cf938c5c8335fbb...
<!-- homu: {"type":"BuildStarted","head_sha":"8d3cf9237dcc9d4f2a7c0bde42dc133e60f51c7d","merge_sha":"52d6d94eb07980426e50416d3cf938c5c8335fbb"} -->

rust-log-analyzer commented:
The job `mingw-check` of your PR [failed](https://github.com/rust-lang-ci/rust/runs/1049583785) ([pretty log](https://rust-lang.github.io/rust-log-analyzer/log-viewer/#https://github.com/rust-lang-ci/rust/commit/52d6d94eb07980426e50416d3cf938c5c8335fbb/checks/1049583785/logs), [raw log](https://github.com/rust-lang-ci/rust/commit/52d6d94eb07980426e50416d3cf938c5c8335fbb/checks/1049583785/logs)). Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.

<details><summary><i>Click to expand the log.</i></summary>

```plain
Build completed successfully in 0:02:12
+ /scripts/validate-toolstate.sh
Cloning into 'rust-toolstate'...
<Nothing changed>
HTTPError: HTTP Error 403: Forbidden
b'{"message":"API rate limit exceeded for user ID 7378925.","documentation_url":"https://docs.github.com/rest/overview/resources-in-the-rest-api#rate-limiting"}'
Traceback (most recent call last):
  File "../../src/tools/publish_toolstate.py", line 288, in <module>
    validate_maintainers(repo, github_token)
  File "../../src/tools/publish_toolstate.py", line 84, in validate_maintainers
    'Accept': 'application/vnd.github.hellcat-preview+json',
  File "/usr/lib/python3.6/urllib/request.py", line 223, in urlopen
    return opener.open(url, data, timeout)
  File "/usr/lib/python3.6/urllib/request.py", line 532, in open
    response = meth(req, response)
  File "/usr/lib/python3.6/urllib/request.py", line 642, in http_response
    'http', request, response, code, msg, hdrs)
  File "/usr/lib/python3.6/urllib/request.py", line 570, in error
    return self._call_chain(*args)
  File "/usr/lib/python3.6/urllib/request.py", line 504, in _call_chain
    result = func(*args)
  File "/usr/lib/python3.6/urllib/request.py", line 650, in http_error_default
    raise HTTPError(req.full_url, code, msg, hdrs, fp)
urllib.error.HTTPError: HTTP Error 403: Forbidden
  local time: Mon Aug 31 05:18:06 UTC 2020
  network time: Mon, 31 Aug 2020 05:18:06 GMT
== end clock drift check ==
== end clock drift check ==
##[error]Process completed with exit code 1.
Terminate orphan process: pid (5093) (node)
Terminate orphan process: pid (5102) (python)
```

</details><p></p>

[I'm a bot](https://github.com/rust-ops/rust-log-analyzer)! I can only do what humans tell me to, so if this was not helpful or you have suggestions for improvements, please ping or otherwise contact **`@rust-lang/infra`**. ([Feature Requests](https://github.com/rust-ops/rust-log-analyzer/issues?q=is%3Aopen+is%3Aissue+label%3Afeature-request))

bors commented:
:broken_heart: Test failed - [checks-actions](https://github.com/rust-lang-ci/rust/runs/1049623819)
<!-- homu: {"type":"BuildFailed","builder_url":"https://github.com/rust-lang-ci/rust/runs/1049623819","builder_name":"checks-actions"} -->

sdroege commented:
The CI failure was some kind of network problem:

```
b'{"message":"API rate limit exceeded for user ID 7378925.","documentation_url":"https://docs.github.com/rest/overview/resources-in-the-rest-api#rate-limiting"}'
```

Dylan-DPC commented:
@bors retry

bors commented:
:hourglass: Testing commit 8d3cf9237dcc9d4f2a7c0bde42dc133e60f51c7d with merge b4b76c287992b99f73159f22a1482866237c9ebd...
<!-- homu: {"type":"BuildStarted","head_sha":"8d3cf9237dcc9d4f2a7c0bde42dc133e60f51c7d","merge_sha":"b4b76c287992b99f73159f22a1482866237c9ebd"} -->

bors commented:
:broken_heart: Test failed - [checks-actions](https://github.com/rust-lang-ci/rust/runs/1051345077)
<!-- homu: {"type":"BuildFailed","builder_url":"https://github.com/rust-lang-ci/rust/runs/1051345077","builder_name":"checks-actions"} -->

sdroege commented:
Also looks like an unrelated failure

nagisa commented:
@bors retry

bors commented:
:hourglass: Testing commit 8d3cf9237dcc9d4f2a7c0bde42dc133e60f51c7d with merge 897ef3a0ec001b89fffe7125c20d6a6bb12fee6c...
<!-- homu: {"type":"BuildStarted","head_sha":"8d3cf9237dcc9d4f2a7c0bde42dc133e60f51c7d","merge_sha":"897ef3a0ec001b89fffe7125c20d6a6bb12fee6c"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/runs/1052300701), [checks-azure](https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_build/results?buildId=36335)
Approved by: nagisa
Pushing 897ef3a0ec001b89fffe7125c20d6a6bb12fee6c to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"nagisa","base_ref":"master","builders":{"checks-azure":"https://dev.azure.com/rust-lang/e71b0ddf-dd27-435a-873c-e30f86eea377/_build/results?buildId=36335","checks-actions":"https://github.com/rust-lang-ci/rust/runs/1052300701"},"merge_sha":"897ef3a0ec001b89fffe7125c20d6a6bb12fee6c"} -->
